sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala

```diff
@@ -296,6 +296,44 @@ case class UpdateTable(
   override def children: Seq[LogicalPlan] = table :: Nil
 }
 
+/**
+ * The logical plan of the MERGE INTO command that works for v2 tables.
+ */
+case class MergeIntoTable(
+    targetTable: LogicalPlan,
+    sourceTable: LogicalPlan,
+    mergeCondition: Expression,
+    matchedActions: Seq[MergeAction],
+    notMatchedActions: Seq[MergeAction]) extends Command with SupportsSubquery {
+  override def children: Seq[LogicalPlan] =
+    Seq(targetTable, sourceTable) ++ Nil
+  override def output: Seq[Attribute] = Seq.empty
+  override lazy val resolved: Boolean = expressions.forall(_.resolved) && childrenResolved &&
+      matchedActions.forall(_.resolved) && notMatchedActions.forall(_.resolved)
+}
+
+sealed abstract class MergeAction(
```

Comment:
We can make it an `Expression`, then everything can be analyzed automatically

Reply:
Done.

Reply:
I found we can't left it be resolved automatically because there're some cases we have to deal with manually like expand stars. It's not easy to extract an expand-star logic in `MergeIntoTable`, so in the latest code I left the reference resolve logic for `MergeIntoTable`.

Reply:
ah i see, then we don't need to make `MergeAction` an `Expression`

Reply:
However, I keep it an `Expression` thus we can leverage the `CheckAnalysis` rules. In the earlier implementation where `MergeAction` is not an `Expression`, I wrote additional check rules in `CheckAnalysis`.  After compared the two ways, I prefer the former.

Reply:
SGTM